File: 14-TypeScript/Grey Multi-Tenant/packages/grey-php/src/Domain/AuthClient.php
<?php

declare(strict_types=1);

namespace Grey\Sdk\Domain;

use Grey\Sdk\Error\GreyError;
use Grey\Sdk\Error\Result;
use Grey\Sdk\Http\HttpClient;

/**
 * Auth domain client for authentication operations.
 */
final class AuthClient
{
    public function __construct(
        private readonly HttpClient $http,
    ) {}

    /**
     * Authenticates a user with email and password.
     *
     * On success, stores the access token in the HTTP client.
     *
     * @return Result<array{accessToken: string, refreshToken: string, expiresIn: int}>
     */
    public function login(string $email, string $password): Result
    {
        $validation = $this->validateCredentials($email, $password);
        if ($validation->isErr()) {
            return $validation;
        }

        $result = $this->http->post('/api/auth/login', [
            'email' => $email,
            'password' => $password,
        ], false);

        if ($result->isOk()) {
            $data = $result->getData();
            $accessToken = $data['accessToken'] ?? $data['access_token'] ?? null;
            if ($accessToken !== null) {
                $this->http->setAccessToken($accessToken);
            }
        }

        return $result;
    }

    /**
     * Logs out the current user.
     *
     * Clears the access token from the HTTP client.
     *
     * @return Result<null>
     */
    public function logout(): Result
    {
        if (!$this->http->isAuthenticated()) {
            return Result::err(GreyError::unauthorized('Not authenticated'));
        }

        $result = $this->http->post('/api/auth/logout');

        // Clear token regardless of result
        $this->http->setAccessToken(null);

        if ($result->isErr()) {
            // Even if logout fails on server, we're logged out locally
            return Result::ok(null);
        }

        return Result::ok(null);
    }

    /**
     * Refreshes the authentication tokens.
     *
     * @return Result<array{accessToken: string, refreshToken: string, expiresIn: int}>
     */
    public function refresh(string $refreshToken): Result
    {
        if (empty($refreshToken)) {
            return Result::err(GreyError::validation('Refresh token cannot be empty'));
        }

        $result = $this->http->post('/api/auth/refresh', [
            'refreshToken' => $refreshToken,
        ], false);

        if ($result->isOk()) {
            $data = $result->getData();
            $accessToken = $data['accessToken'] ?? $data['access_token'] ?? null;
            if ($accessToken !== null) {
                $this->http->setAccessToken($accessToken);
            }
        }

        return $result;
    }

    /**
     * Validates login credentials.
     */
    private function validateCredentials(string $email, string $password): Result
    {
        if (empty($email)) {
            return Result::err(GreyError::validation('Email cannot be empty'));
        }

        if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
            return Result::err(GreyError::validation('Invalid email format'));
        }

        if (empty($password)) {
            return Result::err(GreyError::validation('Password cannot be empty'));
        }

        return Result::ok(null);
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/domain/auth_client.rb
# frozen_string_literal: true

module GreySdk
  module Domain
    # Authentication client for login, logout, and token refresh
    class AuthClient
      # @param http_client [GreySdk::Http::HttpClient] The HTTP client
      def initialize(http_client)
        @http = http_client
      end

      # Authenticates a user with email and password
      # @param email [String] User email
      # @param password [String] User password
      # @return [GreySdk::Error::Result] Result containing tokens or error
      def login(email:, password:)
        return validation_error("Email is required") if email.nil? || email.empty?
        return validation_error("Password is required") if password.nil? || password.empty?

        @http.post("/api/v1/auth/login", {
          email: email,
          password: password
        }).then do |data|
          # Store the access token for subsequent requests
          @http.access_token = data["access_token"] if data["access_token"]
          Error::Result.ok(data)
        end
      end

      # Logs out the current user
      # @return [GreySdk::Error::Result] Result indicating success or error
      def logout
        result = @http.post("/api/v1/auth/logout")
        @http.access_token = nil if result.ok?
        result
      end

      # Refreshes the current access token
      # @param refresh_token [String] The refresh token
      # @return [GreySdk::Error::Result] Result containing new tokens or error
      def refresh(refresh_token:)
        return validation_error("Refresh token is required") if refresh_token.nil? || refresh_token.empty?

        @http.post("/api/v1/auth/refresh", {
          refresh_token: refresh_token
        }).then do |data|
          @http.access_token = data["access_token"] if data["access_token"]
          Error::Result.ok(data)
        end
      end

      # Sets the access token manually
      # @param token [String, nil] Access token
      def set_token(token)
        @http.access_token = token
      end

      private

      def validation_error(message)
        Error::Result.err(Error::GreyError.validation(message))
      end
    end
  end
end
